llama.cpp

```diff
@@ -700,28 +700,31 @@ struct llama_model_loader {
 
         if (use_mmap) {
             mapping.reset(new llama_mmap(&file_loaders.at(0)->file, prefetch_size));
-            if (!lmlock) {
-                // Don't call the callback since the actual loading will be lazy
-                // and we can't measure it.
-                progress_callback = NULL;
-            }
             if (lmlock) {
                 lmlock->init(mapping->addr);
             }
         }
 
         size_t done_size = 0;
         for (llama_load_tensor & lt : tensors_map.tensors) {
-            if (lt.ggml_tensor->backend != GGML_BACKEND_CPU) {
-                continue;
-            }
             if (progress_callback) {
                 progress_callback((float) done_size / data_size, progress_callback_user_data);
             }
             LLAMA_ASSERT(lt.ggml_tensor); // unused tensors should have been caught by load_data already
             lt.data = (uint8_t *) lt.ggml_tensor->data;
             load_data_for(lt);
-            lt.ggml_tensor->data = lt.data;
+            switch(lt.ggml_tensor->backend) {
+                case GGML_BACKEND_CPU:
+                    lt.ggml_tensor->data = lt.data;
+                    break;
+#ifdef GGML_USE_CUBLAS
+                case GGML_BACKEND_CUDA:
+                    ggml_cuda_load_data(lt.data, lt.ggml_tensor);
+                    break;
+#endif
+                default:
+                    continue;
+            }
             done_size += lt.size;
             if (use_mmap && lmlock) {
                 lmlock->grow_to(done_size);
```

Comment:
This will cause the kernel to mlock the parts of the file into memory that are only in VRAM.

Reply:
This cannot avoid if we are use mmap and mlock.

Reply:
You can at least avoid mlock by moving line 725 into the `GGML_BACKEND_CPU` case. I intentionally put the GPU layers at the back of the model so something like this could work.

Reply:
okay